### PR TITLE
Add support for auto_generate_synonyms_phrase_query in match_query, multi_match_query, query_string and simple_query_string

### PR DESCRIPTION
**Pull request by jimczi**
This is a revival of https://github.com/elastic/elasticsearch/pull/23147:
This change adds a new parameter called auto_generate_synonyms_phrase_query (defaults to true).
This option can be used in conjunction with synonym_graph token filter to generate phrase queries
when multi terms synonyms are encountered.
For example, a synonym like "ny, new york" would produce the following boolean query when "ny city" is parsed:
((ny OR "new york") AND city)
Note how the multi terms synonym "new york" produces a phrase query.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java [hunk 1]
```diff
@@ -153,6 +160,9 @@ protected void doWriteTo(StreamOutput out) throws IOException {
         out.writeOptionalString(fuzzyRewrite);
         out.writeOptionalWriteable(fuzziness);
         out.writeOptionalFloat(cutoffFrequency);
+        if (out.getVersion().onOrAfter(Version.V_7_0_0_alpha1)) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java]: are you not going to backport to 6.x?

File: core/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java [hunk 2]
```diff
@@ -238,6 +241,7 @@ public void testFromJson() throws IOException {
                 "    \"max_expansions\" : 50,\n" +
                 "    \"lenient\" : false,\n" +
                 "    \"zero_terms_query\" : \"NONE\",\n" +
+                "      \"auto_generate_synonyms_phrase_query\" : true,\n" +
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java]: indentation looks wrong

**Review by jpountz — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 1 above)
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java]: I will but I can't disable all bwc tests that use one of the query impacted by this change so I'll change the version after the backport.

Comment (weberhofer) [inline core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java]: In 5.x this feature will not be available, will it?

Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java]: Right, it is too late to add this to 5.x so the target for this new feature is 6.1 and 7.

**Comment by weberhofer** [comment]
Thanks for this huge improvement for searching with synonyms!
Short question: Is there a way to see which phrases were used for searching?

**Comment by jpountz** [comment]
maybe you could pass `rewrite=true` to the [validate API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-validate.html) to see what the final query looks like?

**Pull request merged**